SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp

```diff
@@ -238,24 +155,172 @@ void SSpatialOutputLog::PollLogFile(const FString& LogFilePath)
+			if (!FPaths::DirectoryExists(LogDirectory))
 			{
-				PollLogFile(LogFilePath);
-			}, PollTimeInterval, false);
+				UE_LOG(LogSpatialOutputLog, Log, TEXT("Spatial local deployment log directory '%s' does not exist. Will create it."), *LogDirectory);
+
+				if (!FPlatformFileManager::Get().GetPlatformFile().CreateDirectoryTree(*LogDirectory))
+				{
+					UE_LOG(LogSpatialOutputLog, Error, TEXT("Could not create the spatial local deployment log directory. The Spatial Output window will not function."));
+					return;
+				}
+			}
+
+			LogDirectoryChangedDelegate = IDirectoryWatcher::FDirectoryChanged::CreateRaw(this, &SSpatialOutputLog::OnLogDirectoryChanged);
+			DirectoryWatcher->RegisterDirectoryChangedCallback_Handle(LogDirectory, LogDirectoryChangedDelegate, LogDirectoryChangedDelegateHandle, IDirectoryWatcher::WatchOptions::IncludeDirectoryChanges | IDirectoryWatcher::WatchOptions::IgnoreChangesInSubtree);
 		}
 	});
 }
 
-void SSpatialOutputLog::FormatAndPrintRawErrorLine(const FString& LogLine)
+void SSpatialOutputLog::ShutdownLogDirectoryWatcher(const FString& LogDirectory) const
+{
+	AsyncTask(ENamedThreads::GameThread, [LogDirectory, LogDirectoryChangedDelegateHandle = LogDirectoryChangedDelegateHandle]
+    {
+        FDirectoryWatcherModule& DirectoryWatcherModule = FModuleManager::LoadModuleChecked<FDirectoryWatcherModule>(TEXT("DirectoryWatcher"));
+        if (IDirectoryWatcher* DirectoryWatcher = DirectoryWatcherModule.Get())
+        {
+            DirectoryWatcher->UnregisterDirectoryChangedCallback_Handle(LogDirectory, LogDirectoryChangedDelegateHandle);
+        }
+    });
+}
+
+void SSpatialOutputLog::OnLogDirectoryChanged(const TArray<FFileChangeData>& FileChanges)
+{
+	// If this is a new folder creation then switch to watching the log files in that new log folder.
+	for (const FFileChangeData& FileChange : FileChanges)
+	{
+		if (FileChange.Action == FFileChangeData::FCA_Added)
+		{
+			AppendOutput(FString::Printf(TEXT("Switched log directory to '%s'"), *FileChange.Filename), ELogVerbosity::Warning, TEXT("SpatialService"));
+
+			// Now we can start reading log files in the new log folder.
+			for (TUniquePtr<FSpatialLogFileReader>& Reader : LogFileReaders)
+			{
+				Reader->ResetLogDirectory(FileChange.Filename);
+			}
+			return;
+		}
+	}
+}
+
+void SSpatialOutputLog::OnClearLog()
+{
+	// SOutputLog will clear the messages and the SelectedLogCategories.
+	SOutputLog::OnClearLog();
+
+	// Clear the AvailableLogCategories and SelectedLogCategories as we generate many worker categories which are hard to parse.
+	Filter.AvailableLogCategories.Reset();
+	Filter.SelectedLogCategories.Reset();
+}
+
+void SSpatialOutputLog::AppendOutput(const FString& LogMessage, ELogVerbosity::Type Verbosity, const FString& Category)
+{
+	// Serialization must be done on the game thread.
+	AsyncTask(ENamedThreads::GameThread, [this, LogMessage, Verbosity, Category]
+    {
+        Serialize(*LogMessage, Verbosity, FName(*Category));
```

Comment:
What's this look like when both logs are on? Got a screenshot?

Reply:
![image](https://user-images.githubusercontent.com/52573875/82665243-4f871100-9c2b-11ea-9d82-d9c899e995eb.png)

The first warning is generated by code, the "WorkerNannyService" line is the only one coming from the `launch.log`. The rest is the `runtime.log`

Reply:
Sexy